[OSD] Wip log alloc predictor (improvement)

aclamk opened the pull request:
This work is attempt to reduce reallocations that are performed during creation of log message.
It estimates what log size would be at creation and allocates memory accordingly. 
Overall performance improvement is expected.

ORIGINAL:
Every log has 80 chars of preallocated space. 
In 99.8% cases additional std::string needed to be allocated.
For each std::string allocated, 89.6% required resize.
Finally mean of 43 bytes were unused per log allocated space.

NEW
Every log is allocated with previously-predicted size.
In 1.7% cases additional std::string needed to be allocated.
For each std::string allocated 0.3% required resize.
Finally mean of 42 bytes were unused per log allocated space.

dachary commented:
<pre>
[       OK ] Log.ManyGatherLogPrebuf (49 ms)
[ RUN      ] Log.ManyGatherLogPrebufOverflow
</pre>
the make check failure probably means the test core dumped with no trace.

liewegas commented:
Cool!  Can you please squash the 3rd commit into the 2nd?

Also, do you think we should pick a different initial guess than 80?

aclamk commented:
Thanks for your kindness.
Squashed.
Initial guess of 80 chars is selected as backward compatibility. 
Since growing of guessed size is immediate, while reducing is geometric, any size that is less then expected log size is good.
So, reasonable initial values are:
80 - because of backward compatibility
1 - minimal size

liewegas commented:
80 seems fine then :)

liewegas commented:
please rebase?

liewegas commented:
<pre>
  <kind>MismatchedFree</kind>
  <what>Mismatched free() / delete / delete []</what>
  <stack>
    <frame>
      <ip>0x4C2C2BC</ip>
      <obj>/usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so</obj>
      <fn>operator delete(void*)</fn>
    </frame>
    <frame>
      <ip>0x95E9A8</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>ceph::log::Log::flush()</fn>
      <dir>/srv/autobuild-ceph/gitbuilder.git/build/out~/ceph-10.0.0-715-g995637e/src/log</dir>
      <file>Log.cc</file>
      <line>223</line>
    </frame>
    <frame>
      <ip>0x95EB6D</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>ceph::log::Log::entry()</fn>
      <dir>/srv/autobuild-ceph/gitbuilder.git/build/out~/ceph-10.0.0-715-g995637e/src/log</dir>
      <file>Log.cc</file>
      <line>368</line>
    </frame>
    <frame>
      <ip>0x58BD181</ip>
      <obj>/lib/x86_64-linux-gnu/libpthread-2.19.so</obj>
      <fn>start_thread</fn>
      <dir>/build/buildd/eglibc-2.19/nptl</dir>
      <file>pthread_create.c</file>
      <line>312</line>
    </frame>
    <frame>
      <ip>0x75B3EFC</ip>
      <obj>/lib/x86_64-linux-gnu/libc-2.19.so</obj>
      <fn>clone</fn>
      <dir>/build/buildd/eglibc-2.19/misc/../sysdeps/unix/sysv/linux/x86_64</dir>
      <file>clone.S</file>
      <line>111</line>
    </frame>
  </stack>
  <auxwhat>Address 0x8192a00 is 0 bytes inside a block of size 224 alloc'd</auxwhat>
  <stack>
    <frame>
      <ip>0x4C2AB80</ip>
      <obj>/usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so</obj>
      <fn>malloc</fn>
    </frame>
    <frame>
      <ip>0x95E165</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>ceph::log::Log::create_entry(int, int, unsigned long*)</fn>
      <dir>/srv/autobuild-ceph/gitbuilder.git/build/out~/ceph-10.0.0-715-g995637e/src/log</dir>
      <file>Log.cc</file>
      <line>193</line>
    </frame>
    <frame>
      <ip>0xA077A1</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>AdminSocket::register_command(std::string, std::string, AdminSocketHook*, std::string)</fn>
      <dir>/srv/autobuild-ceph/gitbuilder.git/build/out~/ceph-10.0.0-715-g995637e/src/common</dir>
      <file>admin_socket.cc</file>
      <line>423</line>
    </frame>
    <frame>
      <ip>0xA269D1</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>CephContext::CephContext(unsigned int, int)</fn>
      <dir>/srv/autobuild-ceph/gitbuilder.git/build/out~/ceph-10.0.0-715-g995637e/src/common</dir>
      <file>ceph_context.cc</file>
      <line>442</line>
    </frame>
</pre>

aclamk commented:
Hello Sage,

Thank you for finding this problem. I was not aware that such problem is even possible.

I am having difficulty replicating the double-free(or invalid free) behavior that valgrind found.
Is it possible to tell which binary at which scenario caused above problem?

Regards,
Adam

liewegas commented on src/log/Log.cc:
```diff
@@ -184,6 +186,25 @@ Entry *Log::create_entry(int level, int subsys)
+{
+  if (true) {
+    size_t size = __atomic_load_n(expected_size, __ATOMIC_RELAXED);
+    void* ptr=malloc(sizeof(Entry) + size);
```
malloc is used to allocate...

liewegas replied:
and delete is used elsewhere to free.  just switch this to new char[...] and switch the deletes to delete[]?

aclamk commented:
Thank you greatly. Fixed, rebased, pushed. 
It was definately improper to have malloc matched with operator delete. 
But I am afraid it will not fix sporadic problems of valgrind. Let's hope I am wrong.

aclamk commented:
Or - sorry I thought of it too late -  would You like me to write delete operator for Entry that does free (to match with malloc) ?

liewegas commented:
This looks good.. I'll run it through again.  Thanks!

liewegas commented on src/log/Log.cc:
```diff
@@ -184,6 +186,25 @@ Entry *Log::create_entry(int level, int subsys)
+{
+  if (true) {
+    size_t size = __atomic_load_n(expected_size, __ATOMIC_RELAXED);
+    void* ptr=::operator new(sizeof(Entry) + size);
```
void *ptr = ::operator new (...);

tchaikov commented on src/common/dout.h:
```diff
@@ -48,7 +48,8 @@ inline std::ostream& operator<<(std::ostream& out, _bad_endl_use_dendl_t) {
     }									\
-    ceph::log::Entry *_dout_e = cct->_log->create_entry(v, sub);	\
+    static size_t _log_exp_length=80; \
+    ceph::log::Entry *_dout_e = cct->_log->create_entry(v, sub, &_log_exp_length);	\
```
@aclamk i don't quite understand how to choose an optimal number for `m_exp_len`? and how this variable help with the performance? in other words, who will be reading from `m_exp_len`?

tchaikov replied:
ahh, nvm. i understand now.

aclamk replied:
Np :-) sorry I was not present. National holiday.

tchaikov replied:
fwiw, we were having a hard time root-causing a related issue. and came up with a fix: https://github.com/ceph/ceph/pull/11733.
